Add an Error schema

RobNicholsGDS opened the pull request:
The schema defines an error object that contains both information about the overall nature of the error, and an array of details of the specific errors.

The design resulted from a review of the following error object descriptions:

- https://cloud.google.com/storage/docs/json_api/v1/status-codes#401-unauthorized
- https://jsonapi.org/format/#error-objects
- https://www.mscharhag.com/api-design/rest-error-format

A typical error could look like this:

```json
{
  "message": "Validation failure",
  "code": "401",
  "errors": [
    {
      "type": "http://example.com/docs/errors/exceeded-max-length",
      "detail": "The attribute 'summary' is too long. It should be less than 250 characters",
      "instance": "identifier:f58ce342-d3cc-4a13-bba4-ac958c39397b",
      "location": "/summary"
    },
    {
      "type": "http://example.com/docs/errors/type-integer",
      "detail": "The distribution byteSize of '1Mb' is not an integer",
      "instance": "identifier:f58ce342-d3cc-4a13-bba4-ac958c39397b",
      "location": "/distribution/byteSize"
    }
  ]
}
```

asmith-nhsx commented:
This looks good, couple of points:
* should this include a `severity` field?
* is there any value in making the contained array of errors also be error objects? That way you can nest errors?

RobNicholsGDS commented:
> This looks good, couple of points:
> 
> * should this include a `severity` field?
> * is there any value in making the contained array of errors also be error objects? That way you can nest errors?

Yes - severity makes sense.

Nesting the errors is an interesting idea - I'll need to think through the implications.

RobNicholsGDS commented:
I've added a severity attribute.

I have also refactored the schema so that additional error could be defined within an error - to support nesting of errors.

With this change the following error becomes valid:

```json
{
  "message": "Validation failure",
  "code": "401",
  "errors": [
    {
      "type": "http://example.com/docs/errors/validation",
      "detail": "The item 'Acme Object' has failed validation",
      "instance": "identifier:f58ce342-d3cc-4a13-bba4-ac958c39397b",
      "location": "/instances/[2]",
      "errors": [
        {
          "type": "http://example.com/docs/errors/type-integer",
          "detail": "The distribution byteSize of '1Mb' is not an integer",
          "instance": "identifier:f58ce342-d3cc-4a13-bba4-ac958c39397b",
          "location": "/instances/[2]/distribution/byteSize",
          "severity": "3"
        }
      ]
    }
  ]
}```
